drivers/i3c/i3c_npcx.c

```diff
@@ -0,0 +1,2055 @@
+{
+	inst->MDATACTRL |= (BIT(NPCX_I3C_MDATACTRL_FLUSHTB) | BIT(NPCX_I3C_MDATACTRL_FLUSHFB));
+}
+
+/*
+ * brief:  Send request and check the request is valid
+ *
+ * param[in] inst  Pointer to I3C register.
+ *
+ * return  0, success
+ *         -ETIMEDOUT check MCTRLDONE timeout.
+ *         -ENOSYS    invalid use of request.
+ */
+static inline int npcx_i3c_send_request(struct i3c_reg *inst, uint32_t mctrl_val)
+{
+	inst->MCTRL = mctrl_val;
+
+	if (npcx_i3c_status_wait_clear(inst, NPCX_I3C_MSTATUS_MCTRLDONE) != 0) {
+		return -ETIMEDOUT;
+	}
+
+	/* Check invalid use of request */
+	if (IS_BIT_SET(inst->MERRWARN, NPCX_I3C_MERRWARN_INVERQ)) {
+		LOG_ERR("%s: Invalid request, merrwarn: %#x", __func__, inst->MERRWARN);
+		return -ENOSYS;
+	}
+
+	return 0;
+}
+
+/* Start DAA procedure and continue the DAA with a Repeated START */
+static inline int npcx_i3c_request_daa(struct i3c_reg *inst)
+{
+	uint32_t val = 0;
+	int ret = 0;
+
+	/* Set IBI response NACK while processing DAA */
+	SET_FIELD(val, NPCX_I3C_MCTRL_IBIRESP, MCTRL_IBIRESP_NACK);
+
+	/* Send DAA request */
+	SET_FIELD(val, NPCX_I3C_MCTRL_REQUEST, MCTRL_REQUEST_PROCESSDAA);
+
+	ret = npcx_i3c_send_request(inst, val);
+	if (ret != 0) {
+		LOG_ERR("Request DAA error, %d", ret);
+	}
+
+	return ret;
+}
+
+/* Tell controller to start auto IBI */
+static inline int npcx_i3c_request_auto_ibi(struct i3c_reg *inst)
+{
+	uint32_t val = 0;
+	int ret = 0;
+
+	SET_FIELD(val, NPCX_I3C_MCTRL_IBIRESP, MCTRL_IBIRESP_ACK);
+	SET_FIELD(val, NPCX_I3C_MCTRL_REQUEST, MCTRL_REQUEST_AUTOIBI);
+
+	ret = npcx_i3c_send_request(inst, val);
+	if (ret != 0) {
+		LOG_ERR("Request auto ibi error, %d", ret);
+	}
+
+	return ret;
+}
+
+/*
+ * brief:  Controller emit start and send address
+ *
+ * param[in] inst     Pointer to I3C register.
+ * param[in] addr     Dyamic address for xfer or 0x7E for CCC command.
+ * param[in] op_type  Request type.
+ * param[in] is_read  Read(true) or write(false) operation.
+ * param[in] read_sz  Read size.
+ *
+ * return  0, success
+ *         else, error
+ */
+static int npcx_i3c_request_emit_start(struct i3c_reg *inst, uint8_t addr,
+				       enum npcx_i3c_mctrl_type op_type, bool is_read,
+				       size_t read_sz)
+{
+	uint32_t mctrl = 0;
+	int ret = 0;
+
+	/* Set request and target address*/
+	SET_FIELD(mctrl, NPCX_I3C_MCTRL_REQUEST, MCTRL_REQUEST_EMITSTARTADDR);
+
+	/* Set operation type */
+	SET_FIELD(mctrl, NPCX_I3C_MCTRL_TYPE, op_type);
+
+	/* Set IBI response NACK in emit start */
+	SET_FIELD(mctrl, NPCX_I3C_MCTRL_IBIRESP, MCTRL_IBIRESP_NACK);
+
+	/* Set dynamic address */
+	SET_FIELD(mctrl, NPCX_I3C_MCTRL_ADDR, addr);
+
+	/* Set read(1) or write(0) */
+	if (is_read == true) {
+		mctrl |= BIT(NPCX_I3C_MCTRL_DIR);
+		SET_FIELD(mctrl, NPCX_I3C_MCTRL_RDTERM, read_sz); /* Set read length */
+	} else {
+		mctrl &= ~BIT(NPCX_I3C_MCTRL_DIR);
+	}
+
+	ret = npcx_i3c_send_request(inst, mctrl);
+	if (ret != 0) {
+		LOG_ERR("Request start error, %d", ret);
+	}
```

Comment:
return `err`? (and `return 0` at the end)

Reply:
Done